src/userobjects/MoabSkinner.C

```diff
@@ -0,0 +1,1008 @@
+
+        std::vector<moab::EntityHandle> ents = _id_to_elem_handles[next];
+        for (const auto ent : ents)
+          local.insert(ent);
+
+        // Get the libMesh element
+        Elem & elem = mesh().elem_ref(next);
+
+        // How many nearest neighbors (general element)?
+        unsigned int NN = elem.n_neighbors();
+
+        // Loop over neighbors
+        for (unsigned int i = 0; i < NN; i++)
+        {
+          const Elem * nnptr = elem.neighbor_ptr(i);
+          // If on boundary, some may be null ptrs
+          if (nnptr == nullptr)
+            continue;
+
+          dof_id_type idnn = nnptr->id();
+
+          // Select only those that are in the current bin
+          if (elems.find(idnn) != elems.end())
+          {
+            new_neighbors.insert(idnn);
+            // Remove from those still available
+            elems.erase(idnn);
+          }
+        }
+      }
+
+      // Found all the new neighbors, done with current set.
+      neighbors = new_neighbors;
+    }
+
+    // Save this moab range of local neighbors
+    localElems.push_back(local);
+  }
+}
+
+void
+MoabSkinner::reset()
+{
+  if (!_fixed_mesh)
+  {
+    _moab.reset(new moab::Core());
+    skinner.reset(new moab::Skinner(_moab.get()));
+    gtt.reset(new moab::GeomTopoTool(_moab.get()));
+  }
+
+  // Clear entity set maps
+  surfsToVols.clear();
+}
+
+unsigned int
+MoabSkinner::getBin(const unsigned int & iVarBin,
+                       const unsigned int & iDenBin,
+                       const unsigned int & iMat) const
+{
+  return _n_temperature_bins * (_n_density_bins * iMat + iDenBin) + iVarBin;
+}
+
+unsigned int
+MoabSkinner::getMatBin(const unsigned int & iDenBin) const
+{
+  return iDenBin;
+}
+
+void
+MoabSkinner::findSurface(const moab::Range & region,
+                            moab::EntityHandle group,
+                            unsigned int & vol_id,
+                            unsigned int & surf_id,
+                            moab::EntityHandle & volume_set)
+{
+  // Create a volume set
+  vol_id++;
+  createVol(vol_id, volume_set, group);
+
+  // Find surfaces from these regions
+  moab::Range tris;  // The tris of the surfaces
+  moab::Range rtris; // The tris which are reversed with respect to their surfaces
+  skinner->find_skin(0, region, false, tris, &rtris);
+
+  // Create surface sets for the forwards tris
+  VolData vdata = {volume_set, Sense::FORWARDS};
+  createSurfaces(tris, vdata, surf_id);
+
+  // Create surface sets for the reversed tris
+  vdata.sense = Sense::BACKWARDS;
+  createSurfaces(rtris, vdata, surf_id);
+}
+
+void
+MoabSkinner::createSurfaces(moab::Range & faces, VolData & voldata, unsigned int & surf_id)
+{
+  if (faces.empty())
+    return;
+
+  // Loop over the surfaces we have already created
+  for (const auto & surfpair : surfsToVols)
+  {
+    // Local copies of surf/vols
+    moab::EntityHandle surf = surfpair.first;
+    std::vector<VolData> vols = surfpair.second;
```

Comment:
I think this can be a `const` reference to the vector to avoid an unecessary memory allocation.

Reply:
This gets pushed_back to later

Reply:
Whoops, my bad!